src/scheduler.cc

```diff
@@ -267,6 +267,45 @@ Status SchedulerService::TaskInfo(ServerContext* context, const TaskInfoRequest*
   return Status::OK;
 }
 
+Status SchedulerService::KillWorkers(ServerContext* context, const KillWorkersRequest* request, KillWorkersReply* reply) {
+  // TODO: Update reference counts
+  std::lock_guard<std::mutex>
+    failed_tasks_lock(failed_tasks_lock_),
+    get_queue_lock(get_queue_lock_),
+    computation_graph_lock(computation_graph_lock_),
+    fntable_lock(fntable_lock_),
+    avail_workers_lock(avail_workers_lock_),
+    task_queue_lock(task_queue_lock_),
+    workers_lock(workers_lock_);
+  // TODO: Fault handling... what if a worker refuses to die? We just assume it dies here.
+  while (!workers_.empty()) {
+    ClientContext client_context;
+    DieRequest die_request;
+    DieReply die_reply;
+    workers_.back().worker_stub->Die(&client_context, die_request, &die_reply);
+    workers_.pop_back();
+  }
+  avail_workers_.clear();
+  fntable_.clear();
+  get_queue_.clear();
+  while (!task_queue_.empty()) {
+    const Task& task = computation_graph_.get_task(task_queue_.back());
+    TaskStatus status;
+    status.set_operationid(task_queue_.back());
+    status.set_function_name(status.function_name());
```

Comment:
Does this line (and the next) actually make sense? Shouldn't it be set using info in `task`?